quinn-proto/src/congestion/bbr.rs

```diff
@@ -0,0 +1,672 @@
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use crate::congestion::bbr_bw_estimation::BandwidthEstimation;
+use crate::congestion::PacketNumber;
+
+use super::{Controller, ControllerFactory};
+use crate::congestion::bbr_min_max::MinMax;
+use crate::connection::paths::RttEstimator;
+
+// Constants based on TCP defaults.
+// The minimum CWND to ensure delayed acks don't reduce bandwidth measurements.
+// Does not inflate the pacing rate.
+const kDefaultMinimumCongestionWindow: u64 = 4 * MAX_DATAGRAM_SIZE;
+
+// The gain used for the STARTUP, equal to 2/ln(2).
+const kDefaultHighGain: f32 = 2.885;
+// The newly derived gain for STARTUP, equal to 4 * ln(2)
+const kDerivedHighGain: f32 = 2.773;
+// The newly derived CWND gain for STARTUP, 2.
+const kDerivedHighCWNDGain: f32 = 2.0;
+// The cycle of gains used during the PROBE_BW stage.
+const kPacingGain: [f32; 8] = [1.25, 0.75, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0];
+
+const kStartupGrowthTarget: f32 = 1.25;
+const kRoundTripsWithoutGrowthBeforeExitingStartup: u8 = 3;
+
+// Do not allow initial congestion window to be greater than 200 packets.
+const kMaxInitialCongestionWindow: u64 = 200;
+
+// Do not allow initial congestion window to be smaller than 10 packets.
+const kMinInitialCongestionWindow: u64 = 10;
+
+const probe_rtt_based_on_bdp: bool = true;
+const drain_to_target: bool = true;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum Mode {
+    // Startup phase of the connection.
+    STARTUP,
+    // After achieving the highest possible bandwidth during the startup, lower
+    // the pacing rate in order to drain the queue.
+    DRAIN,
+    // Cruising mode.
+    PROBE_BW,
+    // Temporarily slow down sending in order to empty the buffer and measure
+    // the real minimum RTT.
+    PROBE_RTT,
+}
+
+// Indicates how the congestion control limits the amount of bytes in flight.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum RecoveryState {
+    // Do not limit.
+    NOT_IN_RECOVERY,
+    // Allow an extra outstanding byte for each byte acknowledged.
+    CONSERVATION,
+    // Allow two extra outstanding bytes for each byte acknowledged (slow
+    // start).
+    GROWTH,
+}
+
+impl RecoveryState {
+    pub fn in_recovery(&self) -> bool {
+        !matches!(self, RecoveryState::NOT_IN_RECOVERY)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct AckAggregationState {
+    bbr_max_ack_height: MinMax,
+    bbr_aggregation_epoch_start_time: Option<Instant>,
+    bbr_aggregation_epoch_bytes: u64,
+}
+
+impl AckAggregationState {
+    fn update_ack_aggregation_bytes(
+        &mut self,
+        newly_acked_bytes: u64,
+        now: Instant,
+        round: u64,
+        max_bandwidth: u64,
+    ) -> u64 {
+        // Compute how many bytes are expected to be delivered, assuming max
+        // bandwidth is correct.
+        let expected_bytes_acked = max_bandwidth
+            * (now - self.bbr_aggregation_epoch_start_time.unwrap_or(now)).as_micros() as u64
```

Comment:
you probably want `now.saturating_duration_since(...)` for safety reasons and conciseness

Reply:
fixed, but it doesn't look any more concise to me. can you make sure it's what you intended?